language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,370 @@
+            inferTypeM inTerm
+
+-- | This checks that a newly-introduced typevariable is correctly kinded.
+-- For typebind, correct means that the the RHS is indeed kinded by the declared kind.
+-- For databind, correct means that a fully-applied typeconstructor has kind *.
+-- (b is ty::K = _) => [check| G !- ty :: K]
+-- (b is data (X::K) tyarg1::K1 ... tyargN::KN  = _) => [check| G, X::K, tyarg1::K1...tyargN::KN !- [X tyarg1 ... tyargN] :: *]
+-------------------------------------------------
+-- checkKindBinding(G,b)
+checkKindBinding :: forall e term uni ann.
+                   AsTypeError e term uni ann
+                 => Binding TyName Name uni ann
+                 -> TypeCheckM uni e ()
+checkKindBinding = \case
+    TypeBind _ (TyVarDecl ann _ k) rhs ->
+        checkKindM ann rhs $ void k
+    DatatypeBind _ dt@(Datatype ann tycon tyargs _ _) ->
+        -- the fully-applied datatype's type must be *-kinded
+        let appliedTyCon :: Type TyName uni ann = mkAppliedTyCon dt
+        in withTyVarDecls (tycon:tyargs) . -- tycon&tyargs have to be in scope
+            checkKindM ann appliedTyCon $ Type ()
+    _ -> pure ()
+
+-- | This checks that a newly-introduced variable has declared the right type for its term.
+-- The input must be prior normalized.
+-- (b is t:T = _) => [check| G !- t : T]
+-------------------------------------------------
+-- checkTypeBinding(G,b)
+checkTypeBinding :: (GShow uni, GEq uni, DefaultUni <: uni
+                   , AsTypeError e (Term TyName Name uni ()) uni ann
+                   , AsTypeErrorExt e uni ann)
+                 => Normalized (Binding TyName Name uni ann)
+                 -> TypeCheckM uni e ()
+checkTypeBinding = unNormalized >>> \case
+     TermBind _ _ (VarDecl ann _ ty) rhs -> checkTypeM ann rhs . Normalized $ void ty
+     _ -> pure ()
+
+-- | Normalizes subtypes of a binding if it is safe (*-kinded) and, in case of datatypebind, pass extra checks.
+--  b~>?νb
+-- (b is t::T=_) => [check|G !- T::*]
+-- (b is data X::K tyarg1::K1...tyargN::KN) => {
+--   forall t in subtypes(b) . {
+--         [check|G,X::K,tyarg1::KN...tyargN::KN !- t::*]
+--         codomain(X) == [X tyarg1 ... tyargN]
+--         nonrec(b) => [infer|G,tyarg1::K1...tyargN::KN !- domain(X)]
+-- }
+-------------------------------------------------
+-- normalizeBinding(G,b)  b~>?νb
+normalizeBinding :: forall e term uni ann.
+                   (AsTypeError e term uni ann, AsTypeErrorExt e uni ann, GEq uni)
+                  => Recursivity
+                  -> Binding TyName Name uni ann
+                  -> TypeCheckM uni e (Normalized (Binding TyName Name uni ann))
+-- do nothing: there is no point of normalizing the typebind's rhs  , @effectfully correct?
+normalizeBinding _ b@TypeBind{} = pure $ Normalized b
+-- check & normalize the termbind's declared type
+normalizeBinding _ b@TermBind{} = Normalized <$>
+    bindingSubtypes checkStarNorm b
```

Comment:
The `bindingSubtypes (TermBind name declaredTy rhsTerm)` only refers to the declaredTy, and checks that it `declaredTy::*` and normalizes it. We need this because in the later step of `checkTypeBinding (TermBind name normalizedDeclaredTy rhs)` we have to call `checkTerm rhs normalizedDeclaredTy` and there we need a normalized type. I don't understand why this is wrong. Perhaps you thought that `bindingSubtypes` goes into the `rhsTerm` ?

Reply:
> Perhaps you thought that `bindingSubtypes` goes into the `rhsTerm`?

Yes, I made that mistake, however it's still not clear to me why we need to prenormalize `declaredTy`. 

>  we have to call `checkTerm rhs normalizedDeclaredTy` and there we need a normalized type

Can you normalize the type right before calling `checkTerm`? In that case you wouldn't need to normalize `declaredTy`, unwrap it from `Normalized` and later wrap it back just to call `checkTerm` over the wrapped normalized type.

So I do understand that you kind of normalize the type right before `checkTerm`, but can you just directly do that without unwrapping-wrapping (and thus without putting the normalized type back into the AST)?

I don't have any concrete solution. It does seem like the current structure allows for more code sharing between the recursive and non-recursive cases, but I also feel that this sharing makes the code look weirdly structured.

I could play with the code to see if something else pops up in my head. @michaelpj thoughts?

Reply:
Just to clarify, I'm not saying there's anything semantically wrong with the code. It just feels like there should be a more direct way of doing things. But I'm not confident at all that such a way does exist and does not result in code duplication between the recursive and non-recursive cases.

Reply:
I pushed new code where I don't do normalization at the whole binding level. I am not sure however if the code is more or less readable now.